File: android/app/src/main/java/com/example/peacock/feature/effects/EffectProgram.kt
package com.example.peacock.feature.effects

import com.example.peacock.feature.runtime.GroupState

enum class EffectTarget { ALL, GROUP_1, GROUP_2, GROUP_3, GROUP_4, GROUP_5, GROUP_6, GROUP_7 }
enum class EffectValueType {
    NUMBER,
    BOOLEAN,
    COLOUR,
    TARGET,
    NUMBER_LIST,
    BOOLEAN_LIST,
    COLOUR_LIST,
    TARGET_LIST,
}
enum class EffectGroupProperty { HUE, SATURATION, VALUE, MODE }
enum class ArithmeticOperator { ADD, SUBTRACT, MULTIPLY, DIVIDE, MODULO, POWER, MIN, MAX }
enum class ComparisonOperator { EQ, NEQ, LT, LTE, GT, GTE }
enum class LogicOperator { AND, OR }
enum class EffectSourceKind { BLOCKS, SCRIPT }
enum class RuntimeInputKey {
    SENSOR_ACCEL_X,
    SENSOR_ACCEL_Y,
    SENSOR_ACCEL_Z,
    SENSOR_MOTION,
    SENSOR_SHAKE,
    SENSOR_GYRO_X,
    SENSOR_GYRO_Y,
    SENSOR_GYRO_Z,
    SENSOR_PITCH,
    SENSOR_ROLL,
    SENSOR_YAW,
    SENSOR_LIGHT,
    SENSOR_NEAR,
    SENSOR_HEADING,
    SENSOR_PRESSURE,
    AUDIO_LEVEL,
    AUDIO_PEAK,
    AUDIO_BASS,
    AUDIO_MID,
    AUDIO_TREBLE,
    AUDIO_BEAT,
    AUDIO_BPM,
}

enum class BuiltinFunction {
    ABS,
    MIN,
    MAX,
    CLAMP,
    POWER,
    ROUND,
    FLOOR,
    CEIL,
    SQRT,
    LOG,
    SIN,
    COS,
    RADIANS,
    DEGREES,
    MAP,
    LERP,
    SMOOTHSTEP,
    SMOOTHERSTEP,
    EASE_IN,
    EASE_OUT,
    EASE_IN_OUT,
    SINE_WAVE,
    TRIANGLE_WAVE,
    SAW_WAVE,
    SQUARE_WAVE,
    RANDOM,
    NOISE_1D,
    FBM_NOISE,
    SMOOTH,
    DEADZONE,
    HYSTERESIS,
    PEAK_HOLD,
    DEBOUNCE,
    RISING_EDGE,
    FALLING_EDGE,
    RGB,
    RED,
    GREEN,
    BLUE,
    HUE,
    SATURATION,
    VALUE,
    MIX_RGB,
    MIX_HSV,
    COMPLEMENT,
    ROTATE_HUE,
    ADJUST_SATURATION,
    ADJUST_VALUE,
    PALETTE_COLOUR,
    RANDOM_COLOUR,
    CYCLE,
    BEAT_PHASE,
    BAR_PHASE,
    LIST_LENGTH,
    MIRROR,
    ROTATE_PATTERN,
    CENTER_SPREAD,
    CENTER_CONTRACT,
    CHASE,
    WAVE_PATTERN,
}

data class EffectCompileIssue(
    val code: String,
    val messageZh: String,
    val messageJa: String,
    val sourceId: String = "",
    val quickFixWaitMs: Long? = null,
    val sourceStart: Int? = null,
    val sourceEnd: Int? = null,
) {
    val combinedMessage: String get() = "$messageZh / $messageJa"
}

data class EffectColour(val hue: Int, val saturation: Int, val value: Int)

sealed interface EffectValue {
    data class Number(val value: Double) : EffectValue
    data class Boolean(val value: kotlin.Boolean) : EffectValue
    data class Colour(val value: EffectColour) : EffectValue
    data class Target(val value: EffectTarget) : EffectValue
    data class ListValue(
        val elementType: EffectValueType,
        val values: MutableList<EffectValue>,
    ) : EffectValue
}

sealed interface EffectTargetRef {
    data object All : EffectTargetRef
    data object AllPixels : EffectTargetRef
    data class Group(val oneBasedIndex: EffectExpression) : EffectTargetRef
    data class Pixel(
        val oneBasedGroup: EffectExpression,
        val oneBasedPixel: EffectExpression,
    ) : EffectTargetRef
    data class PixelAt(val oneBasedIndex: EffectExpression) : EffectTargetRef
    data class Value(val expression: EffectExpression) : EffectTargetRef

    companion object {
        fun static(target: EffectTarget): EffectTargetRef =
            if (target == EffectTarget.ALL) All
            else Group(EffectExpression.NumberLiteral(target.ordinal.toDouble()))
    }
}

sealed interface EffectExpression {
    val type: EffectValueType

    data class NumberLiteral(val value: Double) : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class BooleanLiteral(val value: Boolean) : EffectExpression {
        override val type = EffectValueType.BOOLEAN
    }

    data class ColourLiteral(val value: EffectColour) : EffectExpression {
        override val type = EffectValueType.COLOUR
    }

    data class Variable(val id: String, override val type: EffectValueType) : EffectExpression
    data object ElapsedMs : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class GroupValue(
        val group: Int,
        val property: EffectGroupProperty,
    ) : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class DynamicGroupValue(
        val oneBasedIndex: EffectExpression,
        val property: EffectGroupProperty,
    ) : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class Arithmetic(
        val operator: ArithmeticOperator,
        val left: EffectExpression,
        val right: EffectExpression,
    ) : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class Clamp(
        val value: EffectExpression,
        val low: EffectExpression,
        val high: EffectExpression,
    ) : EffectExpression {
        override val type = EffectValueType.NUMBER
    }

    data class Comparison(
        val operator: ComparisonOperator,
        val left: EffectExpression,
        val right: EffectExpression,
    ) : EffectExpression {
        override val type = EffectValueType.BOOLEAN
    }

    data class Logic(
        val operator: LogicOperator,
        val left: EffectExpression,
        val right: EffectExpression,
    ) : EffectExpression {
        override val type = EffectValueType.BOOLEAN
    }

    data class Not(val value: EffectExpression) : EffectExpression {
        override val type = EffectValueType.BOOLEAN
    }

    data class ColourFromHsv(
        val hue: EffectExpression,
        val saturation: EffectExpression,
        val value: EffectExpression,
    ) : EffectExpression {
        override val type = EffectValueType.COLOUR
    }

    data class TargetLiteral(val target: EffectTarget) : EffectExpression {
        override val type = EffectValueType.TARGET
    }

    data class TargetFromIndex(val oneBasedIndex: EffectExpression) : EffectExpression {
        override val type = EffectValueType.TARGET
    }

    data class RuntimeInput(
        val key: RuntimeInputKey,
        override val type: EffectValueType =
            if (key == RuntimeInputKey.AUDIO_BEAT) EffectValueType.BOOLEAN else EffectValueType.NUMBER,
    ) : EffectExpression

    data class Builtin(
        val function: BuiltinFunction,
        val arguments: List<EffectExpression>,
        override val type: EffectValueType,
        val nodeId: String = "",
    ) : EffectExpression

    data class ListLiteral(
        val elements: List<EffectExpression>,
        override val type: EffectValueType,
    ) : EffectExpression

    data class ListGet(
        val list: EffectExpression,
        val index: EffectExpression,
        override val type: EffectValueType,
    ) : EffectExpression

    data class FunctionCall(
        val name: String,
        val arguments: List<EffectExpression>,
        override val type: EffectValueType,
        val nodeId: String = "",
    ) : EffectExpression
}

data class EffectFunctionParameter(
    val name: String,
    val variableId: String,
    val type: EffectValueType,
)

data class EffectFunctionDefinition(
    val name: String,
    val parameters: List<EffectFunctionParameter>,
    val returnType: EffectValueType?,
    val operations: List<EffectOp>,
    val returnExpression: EffectExpression? = null,
    val localVariableIds: Set<String> = emptySet(),
)

sealed interface EffectOp {
    val blockId: String

    data class SetHsv(
        val target: EffectTargetRef,
        val h: EffectExpression,
        val s: EffectExpression,
        val v: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(target: EffectTarget, h: Int, s: Int, v: Int, blockId: String = "") :
            this(EffectTargetRef.static(target), EffectExpression.NumberLiteral(h.toDouble()),
                EffectExpression.NumberLiteral(s.toDouble()), EffectExpression.NumberLiteral(v.toDouble()), blockId)
        constructor(
            target: EffectTarget,
            h: EffectExpression,
            s: EffectExpression,
            v: EffectExpression,
            blockId: String = "",
        ) : this(EffectTargetRef.static(target), h, s, v, blockId)
    }

    data class SetColour(
        val target: EffectTargetRef,
        val colour: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(target: EffectTarget, colour: EffectExpression, blockId: String = "") :
            this(EffectTargetRef.static(target), colour, blockId)
    }

    data class FadeHsv(
        val target: EffectTargetRef,
        val h: EffectExpression,
        val s: EffectExpression,
        val v: EffectExpression,
        val durationMs: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(target: EffectTarget, h: Int, s: Int, v: Int, durationMs: Long, blockId: String = "") :
            this(EffectTargetRef.static(target), EffectExpression.NumberLiteral(h.toDouble()), EffectExpression.NumberLiteral(s.toDouble()),
                EffectExpression.NumberLiteral(v.toDouble()), EffectExpression.NumberLiteral(durationMs.toDouble()), blockId)
        constructor(
            target: EffectTarget,
            h: EffectExpression,
            s: EffectExpression,
            v: EffectExpression,
            durationMs: EffectExpression,
            blockId: String = "",
        ) : this(EffectTargetRef.static(target), h, s, v, durationMs, blockId)
    }

    data class FadeColour(
        val target: EffectTargetRef,
        val colour: EffectExpression,
        val durationMs: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(
            target: EffectTarget,
            colour: EffectExpression,
            durationMs: EffectExpression,
            blockId: String = "",
        ) : this(EffectTargetRef.static(target), colour, durationMs, blockId)
    }

    data class AdjustHsv(
        val target: EffectTargetRef,
        val dh: EffectExpression,
        val ds: EffectExpression,
        val dv: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(target: EffectTarget, dh: Int, ds: Int, dv: Int, blockId: String = "") :
            this(EffectTargetRef.static(target), EffectExpression.NumberLiteral(dh.toDouble()),
                EffectExpression.NumberLiteral(ds.toDouble()), EffectExpression.NumberLiteral(dv.toDouble()), blockId)
        constructor(
            target: EffectTarget,
            dh: EffectExpression,
            ds: EffectExpression,
            dv: EffectExpression,
            blockId: String = "",
        ) : this(EffectTargetRef.static(target), dh, ds, dv, blockId)
    }

    data class SetMode(
        val target: EffectTargetRef,
        val mode: EffectExpression,
        val param: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(target: EffectTarget, mode: Int, param: Int, blockId: String = "") :
            this(EffectTargetRef.static(target), EffectExpression.NumberLiteral(mode.toDouble()),
                EffectExpression.NumberLiteral(param.toDouble()), blockId)
        constructor(
            target: EffectTarget,
            mode: EffectExpression,
            param: EffectExpression,
            blockId: String = "",
        ) : this(EffectTargetRef.static(target), mode, param, blockId)
    }

    data class Wait(
        val durationMs: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp {
        constructor(durationMs: Long, blockId: String = "") :
            this(EffectExpression.NumberLiteral(durationMs.toDouble()), blockId)
    }

    data class SetVariable(
        val id: String,
        val value: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp

    data class ChangeVariable(
        val id: String,
        val delta: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp

    data class SetListItem(
        val id: String,
        val index: EffectExpression,
        val value: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp

    data class SeedRandom(
        val seed: EffectExpression,
        override val blockId: String = "",
    ) : EffectOp

    data class CallFunction(
        val name: String,
        val arguments: List<EffectExpression>,
        override val blockId: String = "",
    ) : EffectOp

    data class If(
        val condition: EffectExpression,
        val thenBody: List<EffectOp>,
        val elseBody: List<EffectOp>,
        override val blockId: String = "",
    ) : EffectOp

    data class Repeat(
        val count: EffectExpression?,
        val body: List<EffectOp>,
        override val blockId: String = "",
    ) : EffectOp

    data class For(
        val variableId: String,
        val from: EffectExpression,
        val to: EffectExpression,
        val step: EffectExpression,
        val body: List<EffectOp>,
        override val blockId: String = "",
    ) : EffectOp

    data class While(
        val condition: EffectExpression,
        val body: List<EffectOp>,
        override val blockId: String = "",
    ) : EffectOp

    data class Break(override val blockId: String = "") : EffectOp
    data class Continue(override val blockId: String = "") : EffectOp
    data class End(override val blockId: String = "") : EffectOp
}

data class CompiledEffect(
    val operations: List<EffectOp>,
    val blockCount: Int,
    val estimatedDurationMs: Long?,
    val astSha256: String,
    val variables: Map<String, EffectValueType> = emptyMap(),
    val requiredInputs: Set<RuntimeInputKey> = emptySet(),
    val randomSeed: Long = 0L,
    val functions: Map<String, EffectFunctionDefinition> = emptyMap(),
    val requiresPixelEffect: Boolean = false,
)

data class EffectProgram(
    val id: String,
    val nameZh: String,
    val nameJa: String,
    val workspaceJson: String,
    val astJson: String,
    val astSha256: String,
    val blockCount: Int,
    val estimatedDurationMs: Long?,
    val createdAt: Long,
    val updatedAt: Long,
    val editorSchema: Int = EffectProgramSchemas.EDITOR,
    val programSchema: Int = EffectProgramSchemas.PROGRAM,
    val sourceKind: EffectSourceKind = EffectSourceKind.BLOCKS,
    val scriptSource: String = "",
)

data class EffectFrame(
    val groups: List<GroupState>,
    val pixels: List<EffectRgb>? = null,
    val finished: Boolean,
    val waiting: Boolean,
    val progress: Float?,
)

data class EffectRgb(val red: Int, val green: Int, val blue: Int) {
    init {
        require(red in 0..255 && green in 0..255 && blue in 0..255)
    }
}
object EffectProgramSchemas {
    const val EDITOR = 4
    const val PROGRAM = 6
}

File: ios/Packages/MauryaEffects/Sources/MauryaEffects/EffectProgram.swift
public enum EffectDiagnosticLanguage: Sendable {
    case simplifiedChinese
    case japanese
}

public struct EffectCompileIssue: Error, Equatable, Sendable {
    public let code: String
    public let messageZh: String
    public let messageJa: String
    public let sourceID: String
    public let quickFixWaitMilliseconds: Int64?
    public let sourceStart: Int?
    public let sourceEnd: Int?

    public init(
        code: String,
        messageZh: String,
        messageJa: String,
        sourceID: String = "",
        quickFixWaitMilliseconds: Int64? = nil,
        sourceStart: Int? = nil,
        sourceEnd: Int? = nil
    ) {
        self.code = code
        self.messageZh = messageZh
        self.messageJa = messageJa
        self.sourceID = sourceID
        self.quickFixWaitMilliseconds = quickFixWaitMilliseconds
        self.sourceStart = sourceStart
        self.sourceEnd = sourceEnd
    }

    public func message(for language: EffectDiagnosticLanguage) -> String {
        switch language {
        case .simplifiedChinese: messageZh
        case .japanese: messageJa
        }
    }
}

public struct EffectCompileError: Error, Equatable, Sendable {
    public let issues: [EffectCompileIssue]

    public init(issues: [EffectCompileIssue]) {
        self.issues = issues
    }
}

public indirect enum EffectTargetReference: Equatable, Sendable {
    case all
    case allPixels
    case group(oneBasedIndex: EffectExpression)
    case pixel(oneBasedGroup: EffectExpression, oneBasedPixel: EffectExpression)
    case pixelAt(oneBasedIndex: EffectExpression)
    case value(EffectExpression)

    public static func fixed(_ target: EffectTarget) -> Self {
        target == .all ? .all : .group(oneBasedIndex: .number(Double(target.groupIndex!)))
    }
}

public indirect enum EffectExpression: Equatable, Sendable {
    case number(Double)
    case boolean(Bool)
    case colour(EffectColour)
    case variable(id: String, type: EffectValueType)
    case elapsedMilliseconds
    case groupValue(zeroBasedGroup: Int, property: EffectGroupProperty)
    case dynamicGroupValue(oneBasedIndex: EffectExpression, property: EffectGroupProperty)
    case arithmetic(ArithmeticOperator, EffectExpression, EffectExpression)
    case clamp(value: EffectExpression, low: EffectExpression, high: EffectExpression)
    case comparison(ComparisonOperator, EffectExpression, EffectExpression)
    case logic(LogicOperator, EffectExpression, EffectExpression)
    case not(EffectExpression)
    case colourFromHSV(hue: EffectExpression, saturation: EffectExpression, value: EffectExpression)
    case target(EffectTarget)
    case targetFromIndex(EffectExpression)
    case runtimeInput(RuntimeInputKey)
    case builtin(BuiltinFunction, arguments: [EffectExpression], type: EffectValueType, nodeID: String = "")
    case list([EffectExpression], type: EffectValueType)
    case listGet(list: EffectExpression, index: EffectExpression, type: EffectValueType)
    case functionCall(name: String, arguments: [EffectExpression], type: EffectValueType, nodeID: String = "")

    public var type: EffectValueType {
        switch self {
        case .number, .elapsedMilliseconds, .groupValue, .dynamicGroupValue, .arithmetic, .clamp:
            .number
        case .boolean, .comparison, .logic, .not:
            .boolean
        case .colour, .colourFromHSV:
            .colour
        case .target, .targetFromIndex:
            .target
        case let .variable(_, type), let .builtin(_, _, type, _), let .list(_, type),
            let .listGet(_, _, type), let .functionCall(_, _, type, _):
            type
        case let .runtimeInput(key):
            key.valueType
        }
    }
}

public struct EffectFunctionParameter: Equatable, Sendable {
    public let name: String
    public let variableID: String
    public let type: EffectValueType

    public init(name: String, variableID: String, type: EffectValueType) {
        self.name = name
        self.variableID = variableID
        self.type = type
    }
}

public struct EffectFunctionDefinition: Equatable, Sendable {
    public let name: String
    public let parameters: [EffectFunctionParameter]
    public let returnType: EffectValueType?
    public let operations: [EffectOperation]
    public let returnExpression: EffectExpression?
    public let localVariableIDs: Set<String>

    public init(
        name: String,
        parameters: [EffectFunctionParameter] = [],
        returnType: EffectValueType? = nil,
        operations: [EffectOperation],
        returnExpression: EffectExpression? = nil,
        localVariableIDs: Set<String> = []
    ) {
        self.name = name
        self.parameters = parameters
        self.returnType = returnType
        self.operations = operations
        self.returnExpression = returnExpression
        self.localVariableIDs = localVariableIDs
    }
}

public indirect enum EffectOperation: Equatable, Sendable {
    case setHSV(EffectTargetReference, h: EffectExpression, s: EffectExpression, v: EffectExpression, blockID: String = "")
    case setColour(EffectTargetReference, EffectExpression, blockID: String = "")
    case fadeHSV(
        EffectTargetReference, h: EffectExpression, s: EffectExpression, v: EffectExpression, duration: EffectExpression,
        blockID: String = "")
    case fadeColour(EffectTargetReference, colour: EffectExpression, duration: EffectExpression, blockID: String = "")
    case adjustHSV(EffectTargetReference, dh: EffectExpression, ds: EffectExpression, dv: EffectExpression, blockID: String = "")
    case setMode(EffectTargetReference, mode: EffectExpression, parameter: EffectExpression, blockID: String = "")
    case wait(EffectExpression, blockID: String = "")
    case setVariable(id: String, value: EffectExpression, blockID: String = "")
    case changeVariable(id: String, delta: EffectExpression, blockID: String = "")
    case setListItem(id: String, index: EffectExpression, value: EffectExpression, blockID: String = "")
    case seedRandom(EffectExpression, blockID: String = "")
    case callFunction(name: String, arguments: [EffectExpression], blockID: String = "")
    case ifElse(EffectExpression, then: [EffectOperation], else: [EffectOperation], blockID: String = "")
    case repeatLoop(count: EffectExpression?, body: [EffectOperation], blockID: String = "")
    case forLoop(
        variableID: String, from: EffectExpression, through: EffectExpression, step: EffectExpression, body: [EffectOperation],
        blockID: String = "")
    case whileLoop(EffectExpression, body: [EffectOperation], blockID: String = "")
    case breakLoop(blockID: String = "")
    case continueLoop(blockID: String = "")
    case end(blockID: String = "")
}

public struct CompiledEffect: Equatable, Sendable {
    public let operations: [EffectOperation]
    public let blockCount: Int
    public let estimatedDurationMilliseconds: Int64?
    public let astSHA256: String
    public let variables: [String: EffectValueType]
    public let requiredInputs: Set<RuntimeInputKey>
    public let randomSeed: Int64
    public let functions: [String: EffectFunctionDefinition]
    public let requiresPixelEffect: Bool

    public init(
        operations: [EffectOperation],
        blockCount: Int,
        estimatedDurationMilliseconds: Int64? = nil,
        astSHA256: String = "",
        variables: [String: EffectValueType] = [:],
        requiredInputs: Set<RuntimeInputKey> = [],
        randomSeed: Int64 = 0,
        functions: [String: EffectFunctionDefinition] = [:],
        requiresPixelEffect: Bool = false
    ) {
        self.operations = operations
        self.blockCount = blockCount
        self.estimatedDurationMilliseconds = estimatedDurationMilliseconds
        self.astSHA256 = astSHA256
        self.variables = variables
        self.requiredInputs = requiredInputs
        self.randomSeed = randomSeed
        self.functions = functions
        self.requiresPixelEffect = requiresPixelEffect
    }
}

public struct EffectGroupState: Equatable, Sendable {
    public var innerMode: Int
    public var innerParameter: Int
    public var hue: Int
    public var saturation: Int
    public var value: Int

    public init(innerMode: Int = 1, innerParameter: Int = 0, hue: Int = 0, saturation: Int = 0, value: Int = 0) {
        self.innerMode = innerMode
        self.innerParameter = innerParameter
        self.hue = hue
        self.saturation = saturation
        self.value = value
    }
}

public struct EffectFrame: Equatable, Sendable {
    public let groups: [EffectGroupState]
    public let pixels: [EffectRGB]?
    public let finished: Bool
    public let waiting: Bool
    public let progress: Float?
}

public enum EffectGeometry: Sendable {
    public static let groupCount = 7
    public static let pixelsPerGroup = 6
    public static let pixelCount = 42
}

private extension EffectTarget {
    var groupIndex: Int? {
        switch self {
        case .all: nil
        case .group1: 1
        case .group2: 2
        case .group3: 3
        case .group4: 4
        case .group5: 5
        case .group6: 6
        case .group7: 7
        }
    }
}
